lib/internal/bootstrap/node.js

```diff
@@ -128,6 +129,64 @@
       setupGlobalConsole();
       setupGlobalURL();
       setupGlobalEncoding();
+
+      let errorCodes;
+      let asyncHooks;
+      // declared separately for name, arrow function to prevent construction
+      const queueMicrotask = (callback) => {
+        if (typeof callback !== 'function') {
+          if (errorCodes === undefined) {
+            errorCodes = NativeModule.require('internal/errors').codes;
+          }
+          throw new errorCodes.ERR_INVALID_ARG_TYPE(
+            'callback', 'function', callback);
+        }
+
+        if (asyncHooks === undefined) {
+          asyncHooks = NativeModule.require('internal/async_hooks');
+        }
+
+        const triggerAsyncId = asyncHooks.getDefaultTriggerAsyncId();
+        const asyncId = asyncHooks.newAsyncId();
+        asyncHooks.emitInit(asyncId,
+                            'Microtask',
+                            triggerAsyncId,
+                            callback);
```

Comment:
I think we should create a `MicroTaskObject`  similar to https://github.com/nodejs/node/blob/44f7c1df89a8cad4f773d80396f023893d60a37c/lib/internal/process/next_tick.js#L73. I would recommend the use of `AsyncResource` instead and avoid calling `emitBefore` directly.

Reply:
We are currently discouraging our users to do the same (those methods are deprecated in the public interface). I think we should stick to the same concept for new APIs.

cc @AndreasMadsen

Reply:
The async_hooks provider API is surprisingly tricky to get exactly correct. The AsyncResource object exists to help ensure it is used correctly and consistently. Definitely think it should be used here for consistency.

Reply:
so i have to call runInAsyncScope every time? will this add a perf hit? i don't want to add reasons for people to not use this.

Reply:
A slight performance hit does not make it any less the right thing to do. People will use it if it's useful and usable and makes sense to use. It won't be any less performant than `process.nextTick()` or `Promise`.